pkg/security/probe/selftests/tester_linux.go

```diff
@@ -26,169 +25,212 @@ import (
 	targetFile, err := os.CreateTemp(tmpDir, "datadog_agent_cws_target_file")
 	if err != nil {
-		return err
+		return "", "", err
 	}
-	t.targetFilePath = targetFile.Name()
 
-	return targetFile.Close()
+	return targetFile.Name(), tmpDir, targetFile.Close()
 }
 
 // RunSelfTest runs the self test and return the result
-func (t *SelfTester) RunSelfTest() ([]string, []string, map[string]*serializers.EventSerializer, error) {
-	if err := t.BeginWaitingForEvent(); err != nil {
-		return nil, nil, nil, fmt.Errorf("failed to run self test: %w", err)
+func (t *SelfTester) RunSelfTest() error {
+	t.Lock()
+	defer t.Unlock()
+
+	t.beginSelfTests()
+
+	for _, selfTest := range t.selfTests {
+		if err := selfTest.GenerateEvent(); err != nil {
+			log.Errorf("self test failed: %s", selfTest.GetRuleDefinition().ID)
+		}
 	}
-	defer t.EndWaitingForEvent()
 
-	t.lastTimestamp = time.Now()
+	return nil
+}
 
-	// launch the self tests
-	var success []string
-	var fails []string
-	testEvents := make(map[string]*serializers.EventSerializer)
+// Start the self tester policy provider
+func (t *SelfTester) Start() {}
 
-	for _, selftest := range FileSelfTests {
-		def := selftest.GetRuleDefinition(t.targetFilePath)
+// WaitForResult wait for self test results
+func (t *SelfTester) WaitForResult(timeout time.Duration, cb func(success []eval.RuleID, fails []eval.RuleID, events map[eval.RuleID]*serializers.EventSerializer)) {
+	timer := time.After(timeout)
 
-		predicate, err := selftest.GenerateEvent(t.targetFilePath)
-		if err != nil {
-			fails = append(fails, def.ID)
-			log.Errorf("Self test failed: %s", def.ID)
-			continue
+	var (
+		success []string
+		fails   []string
+		events  = make(map[eval.RuleID]*serializers.EventSerializer)
+	)
+
+	for {
+	LOOP:
+		for {
+			select {
+			case <-t.done:
+				return
+			case event := <-t.eventChan:
+				t.Lock()
+				for _, selfTest := range t.selfTests {
+					if !selfTest.IsSuccess() {
+						selfTest.HandleEvent(event)
+					}
+
+					if selfTest.IsSuccess() {
+						id := selfTest.GetRuleDefinition().ID
+						events[id] = event.Event
+					}
+				}
+				t.Unlock()
+
+				// all test passed
+				if len(events) == len(t.selfTests) {
+					break LOOP
+				}
+			case <-timer:
+				break LOOP
+			}
 		}
-		event, err2 := t.expectEvent(predicate)
-		testEvents[def.ID] = event
-		if err2 != nil {
-			fails = append(fails, def.ID)
-			log.Errorf("Self test failed: %s", def.ID)
-		} else {
-			success = append(success, def.ID)
+
+		t.Lock()
+		for _, selfTest := range t.selfTests {
+			id := selfTest.GetRuleDefinition().ID
+
+			if _, ok := events[id]; ok {
+				success = append(success, id)
+			} else {
+				fails = append(fails, id)
+			}
 		}
-	}
+		t.success, t.fails, t.lastTimestamp = success, fails, time.Now()
+		t.Unlock()
 
-	// save the results for get status command
-	t.success = success
-	t.fails = fails
+		cb(success, fails, events)
 
-	return success, fails, testEvents, nil
+		t.endSelfTests()
```

Comment:
Shouldn't we break out of the loop here? So that we avoid reporting the result multiple times?

Reply:
we don't want to break as we want to handle self tests through gRPC as well